### ISSUE TITLE
jax-rrtmgp 0.3.0 is ~30% slower than 0.2.1 in every RRTMGP configuration; release-wide throughput down 30-41% vs #638

### ISSUE THREAD

**Issue by duncanwp** [climate-analytics-lab/jax-gcm#779]
## What is missing

Every jcm configuration that uses RRTMGP radiation is **30–41 % slower** than the
runtime-regression baselines recorded in #638, measured on `origin/dev`
`f5c8b354` (the pre-release benchmark sweep, Derecho A100-SXM4-40GB). The two
supported configurations that do **not** call RRTMGP — SPEEDY (grey) and the NN
radiation emulator — are at or above their baselines, which is what localises it.

A controlled A/B at **fixed jcm** shows the cause is the `jax-rrtmgp`
`0.2.1 → 0.3.0` upgrade, not a jcm change.

### Release benchmark sweep — `f5c8b354`, 1 month (30 d), chunk 5 d, f32, 1×A100-40GB

| config | grid | days/hr | s/sim-day | peak GiB | util % | baseline d/hr | vs base |
|---|---|---|---|---|---|---|---|
| speedy-t31 (grey) | T31 L8 | 13846 | 0.26 | 0.70 | 96 | ~12900 (#638) | **+7 %** |
| speedy-t31, 12 mo | T31 L8 | 14118 | 0.26 | 0.57 | 96 | ~12900 (#638) | **+9 %** |
| t63-echam-emulated-2m (NN rad) | T63 L47 | 692.3 | 5.20 | 8.51 | 98 | 705.9 (2026-08-26) | −2 % |
| t63-echam-1m | T63 L47 | 160.9 | 22.37 | 8.54 | 89 | 254 (#638) | **−36 %** |
| t63-echam-2m | T63 L47 | 156.7 | 22.98 | 8.54 | 89 | 158.5 (2026-08-25) | −1 % |
| t63-echam-jam | T63 L47 | 112.1 | 32.12 | 16.57 | 90 | 162 (#638) | **−31 %** |
| ma-t63-l47 | T63 L47 | 111.8 | 32.19 | 16.57 | 91 | 162 (#638) | **−31 %** |
| ma-t63-l95 | T63 L95 | 56.1 | 64.13 | 32.60 | 95 | 80 (#638) | **−30 %** |
| t106-echam-1m | T106 L47 | 72.6 | 49.62 | 16.73 | 95 | ~121 (#638) | **−40 %** |
| t106-echam-2m | T106 L47 | 71.4 | 50.39 | 16.74 | 95 | ~121 (#638) | **−41 %** |

Every entry converged on the last-two-chunk 3 % criterion, completed 30/30 days
with `NaN vars: 0`, and was measured with `tools/benchmark.py` (chunk 1 discarded).

`t63-echam-2m` is the key control: it is **unchanged** against a benchmark taken
on 2026-08-25 (158.5 d/hr), i.e. *after* the library upgrade — so nothing in jcm
since Aug 25 costs anything. The whole drop sits in the 2026-08-16 → 2026-08-25
window, in which `jax-rrtmgp` went 0.2.1 → 0.3.0 (installed 2026-08-22; pinned by
#719 on 2026-08-24).

### The A/B: same card, same jcm, only the library differs

`tools/benchmark.py --preset t63-echam-1m`, both arms back-to-back on GPU 0 of
`deg0049`, jcm `f5c8b354`, f32, 30 days:

| arm | days/hr | s/sim-day | 5-day chunk walls | peak GiB | power W |
|---|---|---|---|---|---|
| jax-rrtmgp **0.3.0** (installed) | 160.0 | 22.50 | 112.4, 112.4, 112.5, 112.5 | 8.54 | 190.8 |
| jax-rrtmgp **0.2.1** (`--pythonpath`) | 228.1 | 15.78 | 78.9, 78.9, 78.9, 78.9 | 8.55 | 171.9 |

**0.3.0 is 29.8 % slower** (0.2.1 is +42.6 % faster); +6.7 s of wall per simulated
day. Provenance from the two `result.json`s confirms the arms really differ:

```
0.3.0: "rrtmgp": {"path": ".../.venvs/jaxgcm/lib/python3.11/site-packages/rrtmgp"}
0.2.1: "rrtmgp": {"path": "/glade/derecho/scratch/duncanwp/tmp/rrtmgp-021/unpacked/rrtmgp"}
       "env": {"PYTHONPATH": "/glade/derecho/scratch/duncanwp/tmp/rrtmgp-021/unpacked:..."}
```

Both arms: exit 0, 30/30 days, `nan_any=false`, converged. 228.1 vs the #638
baseline of 254 leaves ~−10 % of jcm-side drift across the same three weeks,
which is inside the 15 % regression threshold and is not flagged here.

## Where the hooks already are

- Baselines: #638 ("Single-GPU rates for the runtime-regression baseline: t63
  ~252 (macsp) / 162 (JAM-L47), t106 ~121, JAM-L95 ~80, SPEEDY ~12,900").
- The pin: `requirements.txt` `jax-rrtmgp>=0.3.0`, raised in `0a2add6d` (#719)
  **for the #711 TOA-flux correctness fix** — so *downgrading is not an available
  remedy*. Either the cost is accepted and the baselines are re-recorded, or the
  0.3.0 hot path is optimised upstream.
- Harness: `tools/benchmark.py` (`--pythonpath` A/B), `tools/profile_terms.py`.
- Raw logs and `result.json`s: `/glade/derecho/scratch/duncanwp/benchmarks/bench-766/`.

## Where the cost lands, and what it is not

`tools/profile_terms.py` at HEAD, `t63-echam-1m`, `--steps 10`, same card, once
per library. **Only the radiation term moves; every other component is identical
to two decimal places:**

| component | 0.2.1 ms/step | 0.3.0 ms/step | kernels/step 0.2.1 -> 0.3.0 |
|---|---:|---:|---:|
| **rrtmgp_radiation** | **85.58** | **132.25** (+54.5 %) | **9,794 -> 13,641 (+39 %)** |
| tiedtke_convection | 8.76 | 8.77 | 1652 -> 1652 |
| dynamics | 5.24 | 5.24 | 405 -> 405 |
| tte_tke_vertical_diffusion | 3.96 | 3.96 | 1532 -> 1532 |
| hines_gwd | 2.91 | 2.91 | 505 -> 505 |
| echam_1m_microphysics | 1.64 | 1.64 | 521 -> 521 |
| lott_miller_sso | 1.36 | 1.36 | 389 -> 389 |

Radiation goes from 76.7 % to 83.6 % of device time. +46.7 ms/step is the entire
regression.

**It is not dtype promotion.** The traced post-optimisation HLO of the whole step
contains **zero `f64` instructions in either arm** (`f64[` count 0/0; everything
in the radiation scope is `f32[`). Nor is it extra transcendental math — 0.3.0
actually issues *fewer* of them:

| in `rrtmgp_radiation` scope (after opt.) | 0.2.1 | 0.3.0 |
|---|---:|---:|
| HLO instructions | 76,270 | 46,948 |
| `exponential(` | 781 | 292 |
| `sqrt(` | 209 | 133 |
| `select(` | 968 | 1,029 |
| `f64[` | 0 | 0 |

So 0.3.0 does **less** arithmetic in **39 % more kernels** and takes 54 % longer:
the rewrite trades fused work for many smaller kernels — a launch/bandwidth cost,
not a FLOP cost. (`_hyperbolics` explicitly shares one `sqrt`/`exp` across four
quantities, which is why the exponential count fell.)

### Source diff 0.2.1 -> 0.3.0

`rte/monochromatic_two_stream.py` is rewritten (~555 changed lines) for
differentiability and float32 precision — NaN-free adjoints at the
conservative-scattering limit, eps-scaled guards replacing fixed `1e-2`/`1e-6`
floors — plus a new `smooth_ops.py` (C1 replacements for hard clamps) and the
#711 TOA-flux change. The solver now selects between truncated-series and
exp-based direct forms with `jnp.where`, and the guarded structure is what
multiplies the kernels:

| file | `jnp.where` 0.2.1 -> 0.3.0 |
|---|---|
| `rte/monochromatic_two_stream.py` | **2 -> 22** |
| `optics/gas_optics.py` | 8 -> 9 |
| `optics/cloud_optics.py` | 4 -> 5 |

`_hyperbolics()` returns four quantities each selected between a series and a
direct branch; `_KTAU_SWITCH` adds a second two-branch structure; the SW direct
beam calls `smooth_ops.smooth_minimum` twice (an extra `sqrt` each) where 0.2.1
used a hard clamp; the Planck geometric mean and the `sqrt` guards add
double-`where` pairs.

## Draft text for the upstream jax-rrtmgp issue (NOT filed — for the maintainer)

> **Title:** 0.3.0 two-stream rewrite costs ~30 % of end-to-end model throughput vs 0.2.1
>
> Benchmarking jax-gcm on one A100-40GB at fixed model code, only swapping the
> library: `t63-echam-1m` (T63L47, RRTMGP + MACv2-SP, 30 simulated days, f32)
> runs at **228.1 sim-days/hr on 0.2.1 and 160.0 on 0.3.0 — 29.8 % slower**,
> +6.7 s of wall per simulated day, converged and NaN-free in both arms.
> Radiation is ~87 % of this configuration's step, so this is essentially all in
> the radiation call.
>
> The 0.3.0 changes are wanted (NaN-free adjoints at conservative scattering, the
> #711 TOA-flux fix, eps-scaled guards), so this is not a request to revert. The
> question is whether the same numerics can be had for less: the rewritten
> `rte/monochromatic_two_stream.py` goes from 2 to 22 `jnp.where` sites, and
> `jnp.where` evaluates both branches, so every g-point pays for the truncated
> series *and* the exp-based direct form of all four `_hyperbolics` quantities on
> every call, plus two `smooth_minimum` `sqrt`s in the SW direct beam.
> Candidates: `lax.select` on a per-column-precomputed mask hoisted out of the
> g-point loop, folding the series/direct choice into one branch-free rational
> approximation, or gating the smooth clamps behind a flag that calibration runs
> turn on and forward-only runs leave off.
>
> Reproducer: `python tools/benchmark.py --preset t63-echam-1m --months 1 --gpu 0
> --f32 [--pythonpath <0.2.1 tree>]` in jax-gcm at `f5c8b354`.

## Not blocking the release

Nothing is unstable: every member completed its month NaN-free and the physical
climate is unaffected by this (it is a cost, not a correctness, change — the
correctness went the right way in #711). The decision this issue asks for is
whether v2.1 ships with the new radiation cost and re-recorded baselines, or
waits on an upstream optimisation.
